src/sparseml/exporters/transforms/utils/matching.py

```diff
@@ -0,0 +1,356 @@
+        match = get_structural_matches(..., parent_ops=[
+            ["A", "B"],
+            ["1", "2"],
+            ["q", "w"],
+            ["G", "H"],
+        ])
+        ```
+        And a match is returned, this list be ordered and sized exactly
+        the way the parent_ops list is ordered.
+        i.e. you can destructure it like so
+        ```python
+        a, b = match.parents[0]
+        _1, _2 = match.parents[1]
+        q, w = match.parents[2]
+        g, h = match.parents[3]
+        ```
+        """
+
+        self.children: List[List[Optional[NodeOrInit]]] = []
+        """
+        This is the sequence of children nodes that were matched via the `children_ops`
+        keyword.
+
+        For example, if you pass in:
+        ```python
+        match = get_structural_matches(..., children_ops=[
+            ["A", "B"],
+            ["1", "2"],
+            ["q", "w"],
+            ["G", "H"],
+        ])
+        ```
+        And a match is returned, this list be ordered and sized exactly
+        the way the parent_ops list is ordered.
+        i.e. you can destructure it like so
+        ```python
+        a, b = match.children[0]
+        _1, _2 = match.children[1]
+        q, w = match.children[2]
+        g, h = match.children[3]
+        ```
+        """
+
+
+def _is_optional_node(tag: str) -> Tuple[str, bool]:
+    if tag.startswith("Optional-"):
+        return tag.split("-")[1], True
+    else:
+        return tag, False
+
+
+def _match_structure(
+    graph: ONNXGraph,
+    node: Union[NodeProto, TensorProto],
+    op_type: str,
+    parent_ops: Optional[List[List[str]]] = None,
+    children_ops: Optional[List[List[str]]] = None,
+) -> Optional[MatchResult]:
+    match = MatchResult(node)
+    if not _match_op_type(match, node, op_type):
+        return None
+    if parent_ops and not _match_parents(match, graph, node, parent_ops):
+        return None
+    if children_ops and not _match_children(match, graph, node, children_ops):
+        return None
+    return match
+
+
+def _match_op_type(match: MatchResult, node: NodeOrInit, op_type: str) -> bool:
+    op_type, is_optional = _is_optional_node(op_type)
+
+    if op_type == INITIALIZER_MATCH and not isinstance(node, TensorProto):
+        return False
+
+    if op_type != INITIALIZER_MATCH and not (
+        isinstance(node, NodeProto) and node.op_type == op_type
+    ):
+        if is_optional:
+            # NOTE: this is handled in `_match_children`
+            match.node = None
+        return is_optional
+    return True
+
+
+def _match_parents(
+    match: MatchResult,
+    graph: ONNXGraph,
+    node: Union[NodeProto, TensorProto],
+    parent_ops: List[List[str]],
+) -> bool:
+    if not (isinstance(node, NodeProto) and len(parent_ops) <= len(node.input)):
+        return False
+
+    parents = graph.get_node_parents(node)
+    assert len(parents) == len(node.input)
+    for parent, expected_op_sequence in zip(parents, parent_ops):
```

Comment:
Its actually necessary to not match exactly against length. e.g. "QuantizeLinear" has 3 inputs but with our recursion we only ever match against 1 input.

So the only checks right now are just to ensure that then length is at least the length of the actual parents.

Reply:
ah I see, so the only issue would be if we want to skip over any of the first `k` inputs to match against any later ones, then they'd need to add placeholder `[]`.

like if we want to match to a QuantizeLinear with an initializer zero point for example it would be `[[], [], [INITIALIZER]]`?

Reply:
Yep exactly!